### DISCUSSION TITLE
Can I make a request handler fully sync in fastapi?

### DISCUSSION THREAD

**Discussion by bobfang1992** [fastapi/fastapi#7046] [Questions] [answered]
I have an async http webs service using fastapi. I am running multiple instances of the same service on the server on a different port and I have an nginx server in front so I can utilise them all. I have a particular resource I need to protect that only one client is accessing it.

```
@app.get("/do_something")
async def do_something():
     critical_section_here()
```

I tried to protect this critical section using a file lock like this:
```
@app.get("/do_something")
async def do_something():
    with FileLock("dosomething.lock"):
        critical_section()
```
This will prevent multiple processes to enter the critical section at the same time. But what I found is that this will actually dead lock. Think about the following event:

1. client 1 connected to port 8000 and enter the critical section
2. while client 1  is still using the resource client 2 is routed to the same port 8000 and then it will try to acquire the file lock, it cannot, so it will keep trying and this will block the execution of client 1 and client 1 will never be able to release the filelock and this means not only this process is locked every other server instance will be locked as well.

So in spite of this, I am just wondering if I can make this request handler fully synced so only one client will be in this request handler? I think to the root of the problem here is two or more client can enter the same request handler here and they will try to acquire the lock at the same time.

Thankss!

**Comment by ycd** [comment]
Well, if i understood your problem correctly, you can achieve that by creating a simple decorator

```python
import asyncio
from functools import wraps


def one_can_connect(func):
    connected = False
    @wraps(func)
    async def wrap(*args, **kwargs):
        nonlocal connected

        if connected is True:
            return 

        connected = True
        response = await func(*args, **kwargs)
        connected = False
        return response

    return wrap


app = FastAPI()


@app.get("/dummy")
@one_can_connect
async def dummy():
    await asyncio.sleep(5)
    return {"foo": "bar"}
```

**Comment by bobfang1992** [comment]
@ycd hi, thanks but I need to coordinate across different processes as well, I am running 8 fastapi instances listening to 8 different ports and due to the nature of my application there can only be one process running the `critical_section` part of my task. So I think the filelock for me is a must.

**Comment by raphaelauv** [comment, marked as answer]
You want a shared lock between different process on the same instance.

That is not a fastAPI question it's a IPC ( inter processus comunication ) question.

There is no direct solution inside the Python Standard Library

You could take a look to 

https://stackoverflow.com/questions/1268252/possible-to-share-in-memory-data-between-2-separate-processes
http://semanchuk.com/philip/PythonIpc/

You can probably close this issue

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:
